File: Chapter 12/Dynamic Scale Beta-t-EGARCH/R/Beta_univ_t_Egarch_Logl_fact.R
# Beta-t-EGARCH likelihood together with filtered diagnostic quantities.

script_args <- commandArgs(trailingOnly = FALSE)
script_file <- sub("^--file=", "", script_args[grep("^--file=", script_args)])
script_dir <- if (length(script_file)) dirname(normalizePath(script_file)) else getwd()
source(file.path(script_dir, "Beta_univ_t_Egarch_Logl.R"))

Beta_univ_t_Egarch_Logl_fact <- function(par, y, inf, I) {
  y <- as.numeric(y)
  p <- unpack_parameters(par, I)
  filtered <- beta_t_egarch_filter(par, y, inf, I)
  fit <- y - filtered$res  # Legacy MATLAB output, kept for compatibility.
  log_density_constant <- lgamma((p$df + 1) / 2) - lgamma(p$df / 2) - 0.5 * log(pi * p$df)
  logl <- length(y) * log_density_constant - sum(filtered$lam) - (p$df + 1) / 2 * sum(log1p(filtered$res^2 / p$df))
  list(Logl = -logl, lam = filtered$lam, res = filtered$res, u = filtered$score, Beta = filtered$beta, fit = fit)
}

File: Chapter 12/Dynamic Scale Beta-t-EGARCH/Python/beta_univ_t_egarch_logl_fact.py
"""Beta-t-EGARCH likelihood together with filtered diagnostic quantities."""

import numpy as np
from scipy.special import gammaln

try:
    from .beta_univ_t_egarch_logl import _unpack_parameters, beta_t_egarch_filter
except ImportError:
    from beta_univ_t_egarch_logl import _unpack_parameters, beta_t_egarch_filter


def beta_univ_t_egarch_logl_fact(par, y, inf, I):
    """Return negative log-likelihood, scale path, residuals, scores, Beta, and fit."""
    y = np.asarray(y, dtype=float).reshape(-1)
    _, _, _, _, df, _, _ = _unpack_parameters(par, I)
    lam, res, score, beta = beta_t_egarch_filter(par, y, inf, I)
    fit = y - res  # Legacy MATLAB output, kept for compatibility.
    log_density_constant = gammaln((df + 1) / 2) - gammaln(df / 2) - 0.5 * np.log(np.pi * df)
    log_likelihood = (
        y.size * log_density_constant
        - np.sum(lam)
        - (df + 1) / 2 * np.sum(np.log1p(res ** 2 / df))
    )
    return -log_likelihood, lam, res, score, beta, fit
